### PR TITLE
Development

### PR DESCRIPTION
**Pull request by NamrataSane**

### CODE REVIEW

File: api/index.ts [hunk 1]
```diff
@@ -1,88 +1,65 @@
...
-  const configKeysLength: number = configKeys?.length;
-  for (let i = 0; i < configKeysLength; i += 1) {
-    const key: any = configKeys[i];
-    const value: any = body[key];
-    // body will have the config object
-    if (root_config.SENSITIVE_CONFIG_KEYS.indexOf(key) > -1) {
-      body[key] = decrypt(value, constants.DECRYPTION.password);
+  configKeys.forEach((key: any) => {
+    if (root_config.SENSITIVE_CONFIG_KEYS.includes(key)) {
+      body[key] = decrypt(body[key], constants.DECRYPTION.password);
```
Comment (NamrataSane) [inline api/index.ts:23]: decrypt(body?.[key]

File: api/index.ts [hunk 2]
```diff
@@ -1,88 +1,65 @@
...
-     * */
-    query.limit =
-      query?.limit > constants.FETCH_PRODUCT_LIMIT ?
-        constants.FETCH_PRODUCT_LIMIT
-        : query?.limit;
-    if (query["sku:in"] || query["id:in"])
-      message = await getSelectedProdsAndCats(query, body);
-    else if (query["categories:in"])
+    // Determine request type and process accordingly
+    if (query["sku:in"] || query["id:in"]) {
```
Comment (NamrataSane) [inline api/index.ts:40]: query?.["sku:in"] || query?.["id:in"]

File: api/index.ts [hunk 3]
```diff
@@ -1,88 +1,65 @@
...
+      // Request for selected products or categories
+      message = await getSelectedProductsAndCategories(query, body);
+    } else if (query["categories:in"]) {
```
Comment (NamrataSane) [inline api/index.ts:43]: query?.["categories:in"]

File: api/package-lock.json
Comment (NamrataSane) [inline api/package-lock.json:1]: In all package.json and package-lock.json change version to v1.1.0

File: api/package.json
Comment (NamrataSane) [inline api/package.json:1]: In all package.json and package-lock.json change version to v1.1.0

File: api/root_config/index.ts [hunk 4]
```diff
@@ -1,41 +1,255 @@
...
+          cs_metadata: {
+            multiConfigName: configKey, // The name of the multi-config used
+            isconfigdeleted: true, // Indicates that the configuration is deleted
+          }
+        };
+      });
+      */
+    } else {
+      const extractedProductID =
+        productPayload?.["id:in"]?.split(",").filter((id: any) => id !== "") ||
```
Comment (NamrataSane) [inline api/root_config/index.ts:79]: ?.filter

File: api/root_config/index.ts [hunk 5]
```diff
@@ -1,41 +1,255 @@
...
+        })
+      );
+    }
+  },
+
+  getSelectedProductsandCategories: (data: any, key: any) => {
+    let url: string = root_config.getUrl(key, data?.query);
+    const urlHasQueryParams: boolean = url?.includes("?");
+
+    url += data["id:in"] ?
```
Comment (NamrataSane) [inline api/root_config/index.ts:143]: data?.

File: api/root_config/index.ts [hunk 6]
```diff
@@ -1,41 +1,255 @@
...
+      `${urlHasQueryParams ? "&" : "?"}id:in=${data["id:in"]}`
```
Comment (NamrataSane) [inline api/root_config/index.ts:144]: data?.

File: api/root_config/index.ts [hunk 7]
```diff
@@ -1,41 +1,255 @@
...
+      : `${urlHasQueryParams ? "&" : "?"}sku:in=${data["sku:in"]}`;
```
Comment (NamrataSane) [inline api/root_config/index.ts:145]: data?.

File: api/root_config/index.ts [hunk 8]
```diff
@@ -1,41 +1,255 @@
...
+      : response;
   },
 
-  // eslint-disable-next-line @typescript-eslint/no-unused-vars
-  getSelectedProductandCatUrl: (data: any, key: any) => {
-    // Add the url for fetching the data from selector page to the custom field here
+  filterProductsByCategory: async (data: any, key: any) => {
+    return _makeApiCall({
+      url: `${root_config.getUrl({ query: data?.query }, key)}?categories:in=${
+        data["categories:in"]
```
Comment (NamrataSane) [inline api/root_config/index.ts:248]: data?.

File: api/root_config/utilityFunctions.ts [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+  function flat(r: any, a: any) {
+    const b: any = {};
+    Object.keys(a)?.forEach((k: any) => {
+      if (k !== "subcategories") {
+        b[k] = a[k];
+      }
+    });
+    r.push(b);
+    if (Array.isArray(a?.subcategories)) {
+      b.subcategories = a?.subcategories.map((item: any) => item?.id);
```
Comment (NamrataSane) [inline api/root_config/utilityFunctions.ts:65]: a?.subcategories?.map

File: api/root_config/utilityFunctions.ts [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+      return a?.subcategories.reduce(flat, r);
```
Comment (NamrataSane) [inline api/root_config/utilityFunctions.ts:66]: a?.subcategories?.reduce

File: api/server.ts [hunk 11]
```diff
@@ -25,7 +25,7 @@ app.post("/", async (req: any, res: any) => {
   res.status(response.statusCode).json(response.body);
 });
 
-const port: number = 8080;
+const port: number = 8081;
```
Comment (NamrataSane) [inline api/server.ts:28]: why port no is changed?

File: api/utils/index.ts [hunk 12]
```diff
@@ -0,0 +1,26 @@
+import CryptoJS from "crypto-js";
+import constants from "../constants";
+
+export const _isEmpty: any = (val: any) =>
+  val === undefined ||
+  val === null ||
+  (typeof val === "object" && !Object.keys(val)?.length) ||
+  (typeof val === "string" && !val.trim().length);
```
Comment (NamrataSane) [inline api/utils/index.ts:8]: ?.length

File: ui/src/common/contexts/customFieldExtensionContext.ts [hunk 13]
```diff
@@ -1,16 +1,39 @@
...
+  appSdkInitialized: boolean;
+  advancedConfig: any[];
+  isOldUser: Boolean;
 };
 
-export const CustomFieldExtensionContext =
-  React.createContext<CustomFieldExtensionContextType>({
-    customField: null,
+export const CustomFieldExtensionContext =  React.createContext<CustomFieldExtensionContextType>({
+    // productCustomField: null,
```
Comment (NamrataSane) [inline ui/src/common/contexts/customFieldExtensionContext.ts:22]: remove commented code

File: ui/src/common/providers/CustomFieldExtensionProvider.tsx [hunk 14]
```diff
@@ -1,48 +1,300 @@
...
+            multiConfigName: "legacy_config",
+            isConfigDeleted: false,
+          },
+        }));
+      }
+      // eslint-disable-next-line
+      else {
+        updatedFieldData = fieldData?.data?.map((fieldDataSet: any) => {
+          if (
+            !fieldDataSet?.cs_metadata
```
Comment (NamrataSane) [inline ui/src/common/providers/CustomFieldExtensionProvider.tsx:118]: verify this condition

File: ui/src/common/providers/EntrySidebarExtensionProvider.tsx [hunk 15]
```diff
@@ -2,28 +2,70 @@ import React, { useEffect, useMemo, useState } from "react";
...
+import rootConfig from "../../root_config";
+import useAppSdk from "../hooks/useAppSdk";
+import useError from "../hooks/useError";
+import useAppConfig from "../hooks/useAppConfig";
 
 const EntrySidebarExtensionProvider: React.FC = function ({ children }: any) {
   const [entryData, setEntry] = useState<{ [key: string]: any }>({});
-  const [loading, setLoading] = useState<boolean>(false);
+  const [contentTypeSchema, setContentTypeSchema] = useState<any>({});
+  // const [loading, setLoading] = useState<boolean>(false);
```
Comment (NamrataSane) [inline ui/src/common/providers/EntrySidebarExtensionProvider.tsx:14]: remove commented code

File: ui/src/common/providers/EntrySidebarExtensionProvider.tsx [hunk 16]
```diff
@@ -2,28 +2,70 @@ import React, { useEffect, useMemo, useState } from "react";
...
+    if (!isNull(location) && !isEmpty(appConfig)) {
+      setAppSdkInitialized(true);
+    }
+  }, [location, appConfig]);
 
   useEffect(() => {
     (async () => {
       if (!isEmpty(entryData) || isNull(location)) return;
-      setLoading(true);
+      // setLoading(true);
```
Comment (NamrataSane) [inline ui/src/common/providers/EntrySidebarExtensionProvider.tsx:30]: remove commented code

File: ui/src/common/providers/EntrySidebarExtensionProvider.tsx [hunk 17]
```diff
@@ -2,28 +2,70 @@ import React, { useEffect, useMemo, useState } from "react";
...
       const entry: { [key: string]: any } = await location?.entry?.getData();
+      const contentTypeUid = location?.entry?.content_type?.uid || "";
+      const contentTypeDetails = await appSdk?.stack?.getContentType(
+        contentTypeUid
+      );
       setEntry(entry);
-      setLoading(false);
+      setContentTypeSchema(contentTypeDetails?.content_type?.schema);
+      // setLoading(false);
```
Comment (NamrataSane) [inline ui/src/common/providers/EntrySidebarExtensionProvider.tsx:38]: remove commented code

File: ui/src/common/providers/EntrySidebarExtensionProvider.tsx [hunk 18]
```diff
@@ -2,28 +2,70 @@ import React, { useEffect, useMemo, useState } from "react";
...
+      // setAppSdkInitialized(true);
```
Comment (NamrataSane) [inline ui/src/common/providers/EntrySidebarExtensionProvider.tsx:39]: remove commented code

File: ui/src/common/providers/EntrySidebarExtensionProvider.tsx [hunk 19]
```diff
@@ -2,28 +2,70 @@ import React, { useEffect, useMemo, useState } from "react";
...
     })();
-  }, [entryData, location, setLoading, setEntry]);
+  }, [entryData, location, setEntry]);
+
+  useEffect(() => {
+    // if (!state.appSdkInitialized) return;
```
Comment (NamrataSane) [inline ui/src/common/providers/EntrySidebarExtensionProvider.tsx:44]: remove commented code

File: ui/src/common/utils/index.tsx [hunk 20]
```diff
@@ -296,7 +301,7 @@ const arrangeSelectedIds = (sortedIdsArray: any[], dataArray: any[]) => {
 
   sortedIdsArray?.forEach((mItem: any) => {
     dataArray?.forEach((sItem: any) => {
-      if (Number(sItem) === Number(mItem)) {
+      if (sItem.toString() === mItem.toString()) {
```
Comment (NamrataSane) [inline ui/src/common/utils/index.tsx:304]: if (sItem?.toString() === mItem?.toString()) {

File: ui/src/common/utils/index.tsx [hunk 21]
```diff
@@ -305,6 +310,69 @@ const arrangeSelectedIds = (sortedIdsArray: any[], dataArray: any[]) => {
...
+    isNotMultiConfigAndSaveInConfig,
+    isNotMultiConfigAndSaveInServerConfig,
+  };
+};
+
+const extractFieldsByConfigType = (configScreen: any) => {
+  const multiConfigFields: string[] = [];
+  const singleConfigFields: string[] = [];
+
+  Object.entries(configScreen).forEach(([key, value]: any) => {
```
Comment (NamrataSane) [inline ui/src/common/utils/index.tsx:350]: ?.forEach

File: ui/src/common/utils/index.tsx [hunk 22]
```diff
@@ -305,6 +310,69 @@ const arrangeSelectedIds = (sortedIdsArray: any[], dataArray: any[]) => {
...
+};
+const arrangeList = (
+  sortedIdsArray: any[],
+  dataArray: any[],
+  uniqueKey: string
+) => {
+  const data: any[] = [];
+  sortedIdsArray?.forEach((mItem: any) => {
+    dataArray?.forEach((sItem: any) => {
+      if (sItem && sItem[uniqueKey] === mItem) {
```
Comment (NamrataSane) [inline ui/src/common/utils/index.tsx:368]: if (sItem?.[uniqueKey] === mItem) {

File: ui/src/containers/App/index.tsx [hunk 23]
```diff
@@ -41,69 +40,69 @@ const SidebarExtension = React.lazy(() => import("../SidebarWidget/index"));
 
 function App() {
   //  below function is called for app signing, i.e. for verifying app tokens in ui
-  const [searchParams] = useSearchParams();
-  if (!rootConfig.verifyAppSigning(searchParams.get("app_token"))) {
-    return <div>{localeTexts.appFailedText.signFail}</div>;
-  }
+  // const [searchParams] = useSearchParams();
```
Comment (NamrataSane) [inline ui/src/containers/App/index.tsx:43]: remove commented code

File: ui/src/containers/ConfigScreen/AddMultiConfigNameModal.tsx [hunk 24]
```diff
@@ -0,0 +1,135 @@
...
+  const [isModalOpen, setIsModalOpen] = useState(false);
+  const onInputChange = (e: any) => {
+    const trimmedValue = e?.target?.value?.trim();
+    setEnteredConfigurationName(trimmedValue);
+
+    if (
+      Object.keys(addMultiConfigurationData)?.length
+      && trimmedValue !== "shopifystore"
+    ) {
+      const isDuplicate = Object.keys(addMultiConfigurationData).some(
```
Comment (NamrataSane) [inline ui/src/containers/ConfigScreen/AddMultiConfigNameModal.tsx:42]: ?.some

File: ui/src/containers/ConfigScreen/DeleteModal.tsx [hunk 25]
```diff
@@ -0,0 +1,97 @@
...
+      toastMessage({
+        text: localtext.configPage.multiConfig.deleteModalNameNotPresent,
+      });
+    }
+  };
+
+  return (
+    <>
+      <ModalHeader
+        title="Confirm Deletion"
```
Comment (NamrataSane) [inline ui/src/containers/ConfigScreen/DeleteModal.tsx:58]: add in the localetext

File: ui/src/containers/ConfigScreen/DeleteModal.tsx [hunk 26]
```diff
@@ -0,0 +1,97 @@
...
+            <b className="multiconfig-ellipsis">{props.multiConfigLabelName}</b>
+          </Tooltip>
+          {localtext.configPage.multiConfig.deleteModal.confirMationYesText}
+        </InstructionText>
+
+        <br />
+
+        <TextInput
+          required
+          placeholder="Enter configuration name for confirmation"
```
Comment (NamrataSane) [inline ui/src/containers/ConfigScreen/DeleteModal.tsx:75]: add in the localetext

File: ui/src/containers/CustomField/index.tsx [hunk 27]
```diff
@@ -26,199 +23,147 @@ import {
...
-            rootConfig.arrangeList(
-              selectedIdsArray,
-              res?.data?.data || res?.data?.items,
-              uniqueKey
-            )
-          );
-      }
-    }
-  };
+  //   location: {},
```
Comment (NamrataSane) [inline ui/src/containers/CustomField/index.tsx:55]: remove commented code

File: ui/src/containers/CustomField/index.tsx [hunk 28]
```diff
@@ -26,199 +23,147 @@ import {
...
+            type: `${appName}_${type}`,
+          });
+        else {
+          const data: any[] = [];
+          const keys = config?.custom_keys?.map((i: any) => i?.value);
+          if (selectedItems?.length) {
+            selectedItems.forEach((item: any) => {
+              const obj1: any = {};
+              keys?.forEach((key: any) => {
+                obj1[key] = item[key];
```
Comment (NamrataSane) [inline ui/src/containers/CustomField/index.tsx:119]: item?.[key]

File: ui/src/containers/CustomField/index.tsx [hunk 29]
```diff
@@ -311,22 +251,20 @@ const CustomField: React.FC<Props> = function ({ type }) {
...
-          </Button>
-        </div>
-      )}
+      <div className="field-extension-wrapper">
+        {renderCustomField()}
+        <Button
+          onClick={handleClick}
+          className="add-product-btn"
+          buttonType="control"
+          disabled={isInvalidCredentials.error || loading}
```
Comment (NamrataSane) [inline ui/src/containers/CustomField/index.tsx:260]: isInvalidCredentials?.

File: ui/src/containers/SidebarWidget/index.tsx [hunk 30]
```diff
@@ -11,106 +13,43 @@ import {
 
 import ProductDescription from "./ProductDescription";
 import WarningMessage from "../../components/WarningMessage";
-import { TypeSDKData, TypeWarningtext } from "../../common/types";
+// import { TypeSDKData, TypeWarningtext } from "../../common/types";
```
Comment (NamrataSane) [inline ui/src/containers/SidebarWidget/index.tsx:16]: remove commented code

File: ui/src/root_config/index.tsx [hunk 31]
```diff
@@ -0,0 +1,880 @@
...
+      return false;
+    }
+    return true;
+  }
+  console.error("Token is missing!");
+  return false;
+};
+// Please refer to the doc for getting more information on each ecommerceEnv fields/keys.
+const ecommerceEnv: EcommerceEnv = {
+  REACT_APP_NAME: "CommerceTools",
```
Comment (NamrataSane) [inline ui/src/root_config/index.tsx:71]: don't use commercetools name here.

File: ui/src/root_config/index.tsx
Comment (NamrataSane) [inline ui/src/root_config/index.tsx:1]: Don't use CommerceTools name in code... Use generalised name

File: ui/src/root_config/index.tsx [hunk 32]
```diff
@@ -0,0 +1,880 @@
...
+  if (!validateMultiConfigKeysByApi) {
+    normalInvalidKeys = checkMultiConfigKeys(
+      configurationData?.multi_config_keys
+    );
+    serverNormalInvalidKeys = checkMultiConfigKeys(
+      serverConfiguration?.multi_config_keys
+    );
+  }
+
+  const normalInvalidKeysList = [
```
Comment (NamrataSane) [inline ui/src/root_config/index.tsx:728]: const normalInvalidKeysList = [
    ...Object.entries(normalInvalidKeys)
     ?.filter(([keys]) => keys?.length)
      ?.map(([source, keys]) => ({ source, keys })),
    ...Object.entries(serverNormalInvalidKeys)
      ?.filter(([keys]) => keys?.length)
      ?.map(([source, keys]) => ({ source, keys })),
  ];

File: ui/src/containers/SelectorPage/index.tsx [hunk 33]
```diff
@@ -270,18 +271,46 @@ const SelectorPage: React.FC = function () {
...
+            meta={metaState}
+            updateList={updateList}
+          />
+        </div>
+
         <InfiniteScrollTable
           uniqueKey={rootConfig.ecommerceEnv.UNIQUE_KEY[config?.type]}
           hiddenColumns={hiddenColumns}
           onToggleColumnSelector={onToggleColumnSelector}
+          // tableHeight={580}
```
Comment (NamrataSane) [inline ui/src/containers/SelectorPage/index.tsx:302]: remove commented code

File: ui/src/containers/SelectorPage/index.tsx [hunk 34]
```diff
@@ -300,14 +329,12 @@ const SelectorPage: React.FC = function () {
           itemStatusMap={itemStatus}
           fetchTableData={fetchData}
           totalCounts={totalCounts}
-          loadMoreItems={loadMoreItems}
+          // loadMoreItems={loadMoreItems}
```
Comment (NamrataSane) [inline ui/src/containers/SelectorPage/index.tsx:332]: remove commented code

File: ui/src/containers/ConfigScreen/index.tsx [hunk 35]
```diff
@@ -56,101 +75,263 @@ const ConfigScreen: React.FC = function () {
...
-    const salt = CryptoJS?.lib?.WordArray?.random(128 / 8);
-    const key = CryptoJS?.PBKDF2(pass, salt, {
-      keySize: keySize / 32,
-      iterations,
+    const updateInstallationData = (prevState: any) => ({
+      ...prevState,
+      installationData: {
+        ...prevState.installationData,
+        configuration: {
+          ...prevState.installationData.configuration,
```
Comment (NamrataSane) [inline ui/src/containers/ConfigScreen/index.tsx:112]: installationData?.configuration

File: ui/src/containers/ConfigScreen/index.tsx [hunk 36]
```diff
@@ -56,101 +75,263 @@ const ConfigScreen: React.FC = function () {
...
+          multi_config_keys: {
+            ...prevState.installationData.configuration.multi_config_keys,
```
Comment (NamrataSane) [inline ui/src/containers/ConfigScreen/index.tsx:114]: .installationData?.configuration?.multi_config_keys

File: ui/src/containers/ConfigScreen/index.tsx [hunk 37]
```diff
@@ -56,101 +75,263 @@ const ConfigScreen: React.FC = function () {
...
+            [accordionId]: result.isMultiConfigAndSaveInConfig,
```
Comment (NamrataSane) [inline ui/src/containers/ConfigScreen/index.tsx:115]: result?.isMultiConfigAndSaveInConfig

File: ui/src/containers/ConfigScreen/index.tsx [hunk 38]
```diff
@@ -56,101 +75,263 @@ const ConfigScreen: React.FC = function () {
...
+          },
+        },
+        serverConfiguration: {
+          ...prevState.installationData.serverConfiguration,
```
Comment (NamrataSane) [inline ui/src/containers/ConfigScreen/index.tsx:119]: .installationData?.serverConfiguration

File: ui/src/containers/ConfigScreen/index.tsx [hunk 39]
```diff
@@ -56,101 +75,263 @@ const ConfigScreen: React.FC = function () {
...
+          multi_config_keys: {
+            ...prevState.installationData.serverConfiguration.multi_config_keys,
```
Comment (NamrataSane) [inline ui/src/containers/ConfigScreen/index.tsx:121]: .installationData?.serverConfiguration?.multi_config_keys

File: ui/src/containers/ConfigScreen/index.tsx [hunk 40]
```diff
@@ -56,101 +75,263 @@ const ConfigScreen: React.FC = function () {
...
+            [accordionId]: result.isMultiConfigAndSaveInServerConfig,
```
Comment (NamrataSane) [inline ui/src/containers/ConfigScreen/index.tsx:122]: result?.isMultiConfigAndSaveInServerConfig

File: ui/src/containers/ConfigScreen/index.tsx [hunk 41]
```diff
@@ -56,101 +75,263 @@ const ConfigScreen: React.FC = function () {
...
-      keySize: keySize / 32,
-      iterations,
-    });
+        const isDefaultKeyExist = multiConfigFields?.length
+          ? checkIsDefaultInitial(state?.installationData?.configuration)
+          : false;
+        const isMultiConfigKeysEmpty = multiConfigFields?.length
+          ? Object.keys(
+              state?.installationData?.configuration.multi_config_keys
+            ).length === 0
```
Comment (NamrataSane) [inline ui/src/containers/ConfigScreen/index.tsx:165]: ?.length === 0

File: api/server.ts (hunk 11 above)
Comment (anonyda) [reply api/server.ts:28]: changed it back

**Pull request merged**